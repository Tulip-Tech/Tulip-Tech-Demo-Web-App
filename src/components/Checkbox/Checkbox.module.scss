@use 'src/styles/variables';
@use 'src/styles/theme';

.checkbox {
  display: inline-block;

  label {
    margin-left: calc(#{variables.$base-spacing} / 2);
    font-size: 14px;
    cursor: pointer;

    a {
      color: variables.$white;
      font-weight: var(--body-font-weight-bold);
      text-decoration: none;

      &:hover {
        text-decoration: underline;
      }
    }
  }

  &.error {
    input {
      &:not(:checked) {
        border: 2px solid theme.$input-field-error-color;
      }
    }
  }

  .header {
    display: inline-block;
    width: 100%;
    margin-bottom: 8px;
    font-weight: var(--body-font-weight-bold);
    text-align: left;

    > span {
      float: right;
      color: rgba(variables.$white, 0.7);
      font-weight: normal;
    }
  }

  input {
    display: inline-flex;
    flex-shrink: 0;
    justify-content: center;
    align-items: center;
    width: 20px;
    height: 20px;
    margin: 8px 0;
    vertical-align: middle;
    border-radius: 2px;
    cursor: pointer;
    transition: all 0.3s;
    -webkit-appearance: none;
    appearance: none;

    &::before {
      display: inline-block;
      color: var(--highlight-contrast-color, variables.$black);
      font-family: arial, sans-serif;
      font-weight: 600;
      font-size: 0;
      content: 'L';
    }

    &:disabled {
      opacity: 0.6;
      pointer-events: none;
      + label {
        cursor: default;
        opacity: 0.6;
      }
    }

    &:hover:not(:disabled) {
      transform: scale(1.1);
    }

    &:not(:checked) {
      border: 2px solid theme.$input-resting-border-color;
      &:hover {
        border-color: theme.$input-hover-border-color;
      }
    }

    &:checked {
      background-color: var(--highlight-color, theme.$primary-color);
      &::before {
        font-size: 20px;
        transform: translateY(-3px) translateX(1px) scaleX(-0.95) rotate(-45deg);
      }
    }
  }
}

.row {
  display: flex;
  align-items: center;
}

.helperTextError {
  margin-top: -5px;
  margin-left: 35px;
  color: theme.$text-field-error-color;
  font-size: 10px;
}
