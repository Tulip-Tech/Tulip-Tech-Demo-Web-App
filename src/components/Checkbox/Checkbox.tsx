import React from 'react';
import classNames from 'classnames';
import { useTranslation } from 'react-i18next';

import styles from './Checkbox.module.scss';

import HelperText from '#components/HelperText/HelperText';
import useOpaqueId from '#src/hooks/useOpaqueId';

type Props = {
  label?: string | JSX.Element;
  name: string;
  value?: string;
  checked?: boolean;
  onChange: React.ChangeEventHandler<HTMLInputElement>;
  header?: string;
  error?: boolean;
  helperText?: string;
  disabled?: boolean;
  required?: boolean;
};

const Checkbox: React.FC<Props> = ({ label, name, onChange, header, checked, value, helperText, disabled, error, required }: Props) => {
  const { t } = useTranslation('common');
  const id = useOpaqueId('check-box', name);

  return (
    <div className={classNames(styles.checkbox, { [styles.error]: error })}>
      {header ? (
        <div className={styles.header}>
          {header}
          {!required ? <span>{t('optional')}</span> : null}
        </div>
      ) : null}
      <div className={styles.row}>
        <input name={name} type="checkbox" id={id} value={value} onChange={onChange} checked={checked} aria-required={required} disabled={disabled} />
        <label htmlFor={id}>
          {required ? '* ' : ''}
          {label}
        </label>
      </div>
      <HelperText error={error} className={error ? styles.helperTextError : undefined}>
        {helperText}
      </HelperText>
    </div>
  );
};

export default Checkbox;
