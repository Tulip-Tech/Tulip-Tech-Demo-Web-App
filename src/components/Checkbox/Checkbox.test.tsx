import React from 'react';
import { render } from '@testing-library/react';

import Checkbox from './Checkbox';

describe('<Checkbox>', () => {
  test('renders and matches snapshot', () => {
    const { container } = render(<Checkbox label="label" name="name" value="value" onChange={vi.fn()} />);

    expect(container).toMatchSnapshot();
  });

  test('should set the checked state', () => {
    const { getByLabelText } = render(<Checkbox label="label" name="name" value="value" checked={true} onChange={vi.fn()} />);

    expect(getByLabelText('label')).toBeChecked();
  });
});
