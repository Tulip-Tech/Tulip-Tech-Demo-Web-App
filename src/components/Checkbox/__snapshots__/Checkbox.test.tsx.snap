// Vitest Snapshot v1

exports[`<Checkbox> > renders and matches snapshot 1`] = `
<div>
  <div
    class="_checkbox_531f07"
  >
    <div
      class="_row_531f07"
    >
      <input
        id="check-box_1235_name"
        name="name"
        type="checkbox"
        value="value"
      />
      <label
        for="check-box_1235_name"
      >
        
        label
      </label>
    </div>
  </div>
</div>
`;
